core/io/resource_format_binary.cpp

```diff
@@ -1520,14 +1520,14 @@ void ResourceFormatSaverBinaryInstance::write_variant(FileAccess *f, const Varia
 			List<Variant> keys;
 			d.get_key_list(&keys);
 
-			for (List<Variant>::Element *E = keys.front(); E; E = E->next()) {
```

Comment:
I think part of the optimization idea is to not have to construct these temporary lists everywhere.  In the above example, we're not gaining anything (a temporary list is still constructed and just used for iteration purposes, although it's ranged now).

The idea here, is we shouldn't need to call `d.get_key_list(&keys);`. We should just iterate over `d` using a KeyPair reference.  No heap allocations.

Reply:
Is this implemented? This line doesn't compile:

```cpp
for (KeyValue<Variant, Variant> &E : d) {
```

`this range-based 'for' statement requires a suitable "begin" function and none was found C/C++(2291)`

Reply:
I have no idea, i just found out reduz implemented range iterators a couple days ago.  Doesn't look like it, based on https://github.com/godotengine/godot/pull/50284 though.  Seems a few more data structures still need the support.

Reply:
Yes the idea is to do the conversion to the next syntax everywhere we can, and then keep working on optimizing usage in the templates directly.